### PR TITLE
Feature/feature classifier

### PR DESCRIPTION
**Pull request by ihowell**
Working usefulness functionality. Was actually working before, just wasn't apparent as the bias of the
final convolutional layer was 0 and thus in printing showed 0's. Fixed up the logic to be in function definitions instead of giant lambdas. Also, was able to remove the num_classes and num_features
parameters and infer them from the data and model.

### CODE REVIEW

File: pyroclast/common/adversarial.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import tensorflow as tf
+import numpy as np
+
+
+def fast_gradient_sign_method(feature_fn, classify_fn, x, y, eps, norm):
+    """
```
Comment (ihowell) [inline pyroclast/common/adversarial.py]: Give a description of the method and what it does.

File: pyroclast/common/adversarial.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                         watch_accessed_variables=False) as tape:
+        tape.watch(x)
+        features = feature_fn(x)
+        fc_dot = tf.einsum('ij,ik->ijk', features, y)
+    jacobian = tape.batch_jacobian(fc_dot, x)
+    return jacobian
+
+
+def linear_optimization(jacobian, eps, norm):
+    """
```
Comment (ihowell) [inline pyroclast/common/adversarial.py]: What does this linear optimization optimize?  Jacboian = gradients, eps = epsilon, norm = norm to minimze?

File: pyroclast/common/adversarial.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    Args:
+        jacobian (Tensor): shape [N,F,Cl,...] with gradients of feature/class outer product w.r.t. data
+        eps (float): norm constraint on the delta
+        norm (1, 2, or np.inf): norm choosing constraint set to optimize
+
+    Returns:
+        scaled_perturbation (Tensor): perturbation of shape [N,F,Cl,...] which \
+            maximizes the use of jacobian w.r.t. scaled to eps limit under norm
```
Comment (ihowell) [inline pyroclast/common/adversarial.py]: Maximizes `use` of Jacobian? what does that mean?

File: pyroclast/common/adversarial.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    Args:
+        feature_fn (x -> z): where x has shape [NHWCh]
+        classify_fn (z -> y): where y has shape [NCl] and z identical to feature_fn
+        x (Tensor): data, assumed to be [NHWCh]
+        y (Tensor): labels, assumed int typed
+        eps (float): norm constraint on the delta
+        norm (1, 2, or np.inf): norm choosing constraint set to optimize
+
+    Returns:
+        perturbed_data (Tensor): adversarially perturbed data to minimize \
```
Comment (ihowell) [inline pyroclast/common/adversarial.py]: Do we want to be returning the perturbed data or the perturbation itself? We might want to analyze the perturbation, so I think we should return the perturbation instead of the modified data.

File: pyroclast/common/feature_classifier.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        Args:
+           features (tf.Tensor): Input of shape [batch_size, num_features]
+
+        Returns:
+           logits (tf.Tensor): Output of shape [batch_size, num_classes]
+        """
+        pass
+
+    def usefulness(self, D):
+        """Finds maximal rho such that f is rho-useful over the dataset D for
```
Comment (ihowell) [inline pyroclast/common/feature_classifier.py]: This description is no longer accurate. These docs should be updated.

File: pyroclast/common/adversarial.py (hunk 1 above)
Comment (ihowell) [reply pyroclast/common/adversarial.py]: Also, talk about how it takes advantage of ReLUs so we don't waste any more time.

File: pyroclast/common/feature_classifier.py (hunk 5 above)
Comment (DrKwint) [reply pyroclast/common/feature_classifier.py]: Could you clarify which part isn't accurate anymore? I'm fairly sure nothing essential has changed that requires changing the description.

Comment (ihowell) [reply pyroclast/common/feature_classifier.py]: Well, it doesn't say anything about calculating rho for each entry in the cross product of features and classes except in the return type. The initial description makes it seem like we are calculating rho for a single f, c.

**Comment by ihowell** [comment]
Not that I can approve now, but 👍
